File: test/extensions/dynamic_modules/test_data/go/listener_integration_test/listener_integration_test.go
package main

import (
	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterListenerFilterConfigFactories(map[string]shared.ListenerFilterConfigFactory{
		"write_to_socket": &writeToSocketConfigFactory{},
		"buffer_read":     &bufferReadConfigFactory{},
	})
}

func main() {}

type writeToSocketConfigFactory struct {
	shared.EmptyListenerFilterConfigFactory
}

func (f *writeToSocketConfigFactory) Create(shared.ListenerFilterConfigHandle, []byte) (shared.ListenerFilterFactory, error) {
	return &writeToSocketFactory{}, nil
}

type writeToSocketFactory struct {
	shared.EmptyListenerFilterFactory
}

func (f *writeToSocketFactory) Create(handle shared.ListenerFilterHandle) shared.ListenerFilter {
	return &writeToSocketFilter{handle: handle}
}

type writeToSocketFilter struct {
	handle shared.ListenerFilterHandle
	shared.EmptyListenerFilter
}

func (f *writeToSocketFilter) OnAccept() shared.ListenerFilterStatus {
	if f.handle.GetConnectionStartTimeMs() == 0 {
		panic("expected connection start time")
	}
	if _, _, ok := f.handle.GetRemoteAddress(); !ok {
		panic("expected remote address")
	}
	if _, _, ok := f.handle.GetLocalAddress(); !ok {
		panic("expected local address")
	}
	f.handle.SetRequestedServerName("sdk.listener.test")
	f.handle.SetDetectedTransportProtocol("sdk_listener")
	return shared.ListenerFilterStatusContinue
}

type bufferReadConfigFactory struct {
	shared.EmptyListenerFilterConfigFactory
}

func (f *bufferReadConfigFactory) Create(shared.ListenerFilterConfigHandle, []byte) (shared.ListenerFilterFactory, error) {
	return &bufferReadFactory{}, nil
}

type bufferReadFactory struct {
	shared.EmptyListenerFilterFactory
}

func (f *bufferReadFactory) Create(handle shared.ListenerFilterHandle) shared.ListenerFilter {
	return &bufferReadFilter{handle: handle}
}

type bufferReadFilter struct {
	handle shared.ListenerFilterHandle
	shared.EmptyListenerFilter
}

func (f *bufferReadFilter) OnAccept() shared.ListenerFilterStatus {
	return shared.ListenerFilterStatusStop
}

func (f *bufferReadFilter) OnData(dataLength uint64) shared.ListenerFilterStatus {
	if dataLength < 4 {
		panic("expected at least 4 bytes")
	}
	chunk, ok := f.handle.GetBufferChunk()
	if !ok || chunk.ToString()[:4] != "ping" {
		panic("unexpected buffer contents")
	}
	if f.handle.GetCurrentMaxReadBytes() != 4 {
		panic("unexpected max read bytes")
	}
	return shared.ListenerFilterStatusContinue
}

func (f *bufferReadFilter) MaxReadBytes() uint64 { return 4 }

File: test/extensions/dynamic_modules/test_data/cpp/listener_integration_test.cc
#include <cassert>

#include "source/extensions/dynamic_modules/sdk/cpp/sdk_listener.h"

namespace Envoy {
namespace DynamicModules {

class WriteToSocketFilter : public ListenerFilter {
public:
  explicit WriteToSocketFilter(ListenerFilterHandle& handle) : handle_(handle) {}

  ListenerFilterStatus onAccept() override {
    assert(handle_.getConnectionStartTimeMs() > 0);
    assert(handle_.getRemoteAddress().has_value());
    assert(handle_.getLocalAddress().has_value());
    handle_.setRequestedServerName("sdk.listener.test");
    handle_.setDetectedTransportProtocol("sdk_listener");
    return ListenerFilterStatus::Continue;
  }

private:
  ListenerFilterHandle& handle_;
};

class WriteToSocketFactory : public ListenerFilterFactory {
public:
  std::unique_ptr<ListenerFilter> create(ListenerFilterHandle& handle) override {
    return std::make_unique<WriteToSocketFilter>(handle);
  }
};

class WriteToSocketConfigFactory : public ListenerFilterConfigFactory {
public:
  std::unique_ptr<ListenerFilterFactory> create(ListenerFilterConfigHandle&,
                                                std::string_view) override {
    return std::make_unique<WriteToSocketFactory>();
  }
};

REGISTER_LISTENER_FILTER_CONFIG_FACTORY(WriteToSocketConfigFactory, "write_to_socket");

class BufferReadFilter : public ListenerFilter {
public:
  explicit BufferReadFilter(ListenerFilterHandle& handle) : handle_(handle) {}

  ListenerFilterStatus onAccept() override { return ListenerFilterStatus::StopIteration; }

  ListenerFilterStatus onData(size_t data_length) override {
    assert(data_length >= 4);
    const auto chunk = handle_.getBufferChunk();
    assert(chunk.has_value());
    assert(chunk->toStringView().substr(0, 4) == "ping");
    assert(handle_.currentMaxReadBytes() == 4);
    return ListenerFilterStatus::Continue;
  }

  size_t maxReadBytes() override { return 4; }

private:
  ListenerFilterHandle& handle_;
};

class BufferReadFactory : public ListenerFilterFactory {
public:
  std::unique_ptr<ListenerFilter> create(ListenerFilterHandle& handle) override {
    return std::make_unique<BufferReadFilter>(handle);
  }
};

class BufferReadConfigFactory : public ListenerFilterConfigFactory {
public:
  std::unique_ptr<ListenerFilterFactory> create(ListenerFilterConfigHandle&,
                                                std::string_view) override {
    return std::make_unique<BufferReadFactory>();
  }
};

REGISTER_LISTENER_FILTER_CONFIG_FACTORY(BufferReadConfigFactory, "buffer_read");

} // namespace DynamicModules
} // namespace Envoy
